### PR TITLE
Feature/Implement ranker for handling near by points/places

### PR DESCRIPTION
**Pull request by CodeBear801**
# Issue

https://github.com/Telenav/osrm-backend/issues/238

- Implement `ranker` to rank near by places.  Could based on `GreatCircleDistance` or  `ShortestDistance` via OSRM
- Add dependency of github.com/blevesearch/bleve for geo.Haversin(discussed in https://github.com/Telenav/osrm-backend/issues/185)
- When sending out table request to OSRM, add point count limit to avoid too big request&response
- Implement `iterator` interface and `FindNearByPointIDs` in indexer
- Change `Latitude`,`Longitude` to `Lat`, `Lon`

To do:
- Ranker's interface might be just need one: RankPlaces()
   + One implmentation based on greatCircleDistance and another based on ShortestPath
   + Most of usage situation just need one ranking strategy instead of two
- Use global `location` to replace local one
- Change the name of `Point` to `Place`, because such `Points` are `POI` or the location with specific meaning, not random point on earth
- Re-try web request when fail

## Tasklist

 - [ ] implement code
 - [ ] add tests
 - [ ] review

### CODE REVIEW

File: integration/service/spatialindexer/ranker/ranker_factory.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+	"github.com/Telenav/osrm-backend/integration/oasis/osrmconnector"
+	"github.com/Telenav/osrm-backend/integration/service/spatialindexer"
+)
+
+const (
+	SimpleRanker    = "SimpleRanker"
+	OSRMBasedRanker = "OSRMBasedRanker"
+)
+
+func CreateRanker(rankerType string, oc *osrmconnector.OSRMConnector) spatialindexer.Ranker {
```
Comment (wangyoucao577) [inline integration/service/spatialindexer/ranker/ranker_factory.go:13]: comment

File: integration/service/spatialindexer/ranker/ranker_factory.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package ranker
+
+import (
+	"github.com/Telenav/osrm-backend/integration/oasis/osrmconnector"
+	"github.com/Telenav/osrm-backend/integration/service/spatialindexer"
+)
+
+const (
+	SimpleRanker    = "SimpleRanker"
```
Comment (wangyoucao577) [inline integration/service/spatialindexer/ranker/ranker_factory.go:9]: comment

File: integration/service/spatialindexer/ranker/simple_ranker.go [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import "github.com/Telenav/osrm-backend/integration/service/spatialindexer"
+
+type simpleRanker struct {
+}
+
+func newSimpleRanker() *simpleRanker {
+	return &simpleRanker{}
+}
+
+func (ranker *simpleRanker) RankPointIDsByGreatCircleDistance(center spatialindexer.Location,
```
Comment (wangyoucao577) [inline integration/service/spatialindexer/ranker/simple_ranker.go:12]: comment

File: integration/service/spatialindexer/s2indexer/indexer.go [hunk 4]
```diff
@@ -63,6 +63,37 @@ func (indexer *S2Indexer) Dump(folderPath string) {
...
+				Location: location,
+			}
+		}
+		close(pointsC)
+	}()
+
+	return pointsC
+}
+
+func (indexer *S2Indexer) FindNearByPointIDs(center spatialindexer.Location, radius float64, limitCount int) []spatialindexer.PointInfo {
```
Comment (wangyoucao577) [inline integration/service/spatialindexer/s2indexer/indexer.go:83]: comment

File: integration/service/spatialindexer/s2indexer/indexer.go [hunk 5]
```diff
@@ -73,6 +104,10 @@ func (indexer S2Indexer) getPointIDsByS2CellID(cellid s2.CellID) ([]spatialindex
 	return pointIDs, ok
 }
 
+func (indexer S2Indexer) isInit() bool {
```
Comment (wangyoucao577) [inline integration/service/spatialindexer/s2indexer/indexer.go:107]: isInitialized?

Comment (CodeBear801) [reply integration/service/spatialindexer/s2indexer/indexer.go:107]: Changed to isInitialized

File: integration/service/spatialindexer/ranker/ranker_factory.go (hunk 1 above)
Comment (CodeBear801) [reply integration/service/spatialindexer/ranker/ranker_factory.go:13]: This comment is missed until today, others has already added.
Will try to fix my linter.

**Pull request merged**